contributors/design-proposals/apiserver-count-fix.md

```diff
@@ -0,0 +1,129 @@
+the Endpoints eventually consistent, but the code relies on the
+Endpoints count becoming equal to or greater than masterCount. When the
+apiservers become greater than the masterCount the Endpoints tend to
+flap.
+
+If the number endpoints were scaled down from automation, then the
+Endpoints would never become consistent.
+
+## Proposal
+
+### Create New Reconciler
+
+| Kubernetes Release  | Quality | Description |
+| ------------- | ------------- | ----------- |
+| 1.9           | alpha         | <ul><li>Add a new reconciler</li><li>Add a command-line type `--alpha-apiserver-endpoint-reconciler-type`<ul><li>configmap</li><li>default</li></ul></li></ul>
+| 1.10          | beta          | <ul><li>Turn on the `configmap` type by default</li></ul>
+| 1.11          | stable        | <ul><li>Remove code for old reconciler</li><li>Remove --apiserver-count</li></ul>
+
+The MasterCountEndpointReconciler does not meet the current needs for
+durability of API Endpoint creation, deletion, or failure cases.
+
+Create a new `MasterEndpointConfigMapReconciler` within
+master/controller.go.
+
+Add a `kube-apiserver-config` ConfigMap in the `kube-system`
+namespace. The duration found within the Coordination map would be configurable by
+admins without a recompile. The ConfigMap would include the following:
+
+```go
+KubeApiServerConfigMap{
+  "expiration-duration":        "1m",
+  ... [other flags potentially]
+}
+```
+
+The Coordination ConfigMap would be formed such that:
+
+```go
+CoordinationConfigMap{
+  "[namespace]/[name]": "[KubeAPIServerEndpoint]",
+}
+```
+
+*Key Format*: ip-[IP String Formatted]-[port]
+
+```go
+KubeAPIServerEndpointConfigMap{
+  "master-count":                "0",
+	"ip-2001-4860-4860--8888-443": "time.Time",
+	"ip-192-168-0-3-443":          "time.Time",
+}
+```
+
+***TODO: should it a serialized struct, or the raw time value?***
+
+The reconcile loop will expire endpoints that do not meet the duration.
+On each reconcile loop (the loop runs every 10 seconds currently,
+but interval will be changed to 80% of the `expiration-duration`):
+
+1. GET `kube-apiserver-endpoints` ConfigMap (as endpointMap)
+1. Update the timestamp for the currently running API server
+1. Remove all endpoints where the timestamp is greater than `expiration-duration` from the configMap.
+1. Do a GET on the `kube-apiserver-endpoints`
+1. Construct a PATCH operation with the deleted endpoints
+
+configmap.yml:
+
+```yaml
+kind: ConfigMap
+apiVersion: v1
+metadata:
+  name: kube-apiserver-endpoints
+  namespace: default
+```
+
+### Alternate Proposals
+
+#### Custom Resource Definitions
+
+CRD's were considered for this proposal, but were not proposed due to
+constraints of having CRDs within core has not been clearly defined.
+Layering could also be an issue, so the proposal defined ConfigMaps as
+the current path forward.
```

Comment:
@bgrant0607 when we discussed this before on whether we should be abusing configmaps for election we discussed a CRD as being acceptable for use within the api server.  

I think my preference at this point is not to abuse configmaps for high volume write traffic that is not related directly to the purpose of delivering configuration into a pod.  If we do not want to use the previously proposed mechanism of using storage with TTL (which I still think is the correct approach since it is how everyone else using etcd deals with leases) then we need to make a case that it is ok for a single component (the core kube-apiservers) to constantly write to a configmap, but that other components should not do so / are not affected by it.